File: blooming_directed_graph_agents/trigger_process_agent/objective-c/TriggerProcessAgent.m
// blooming_directed_graph_trigger_process_agent – Objective-C implementation. Namespace: LangFlow.
// IEEE 752 64-bit geometrica Z field (zed): see docs/IEEE_752_64BIT_GEOMETRICA_Z_FIELD_ZED_SPECIFICATIONS.md
#import "TriggerProcessAgent.h"
#import "Collection.h"  // from filter_agents/objective-c; add -I when building
#import <stdlib.h>

@implementation BloomingDirectedGraphTriggerProcessAgent
+ (NSDictionary *)triggerWithNode:(BDGNode *)node action:(NSString *)action payload:(NSDictionary *)payload {
    if ([action isEqualToString:@"workflow_dispatch"]) {
        return @{ @"ok": @(YES) };
    }
    if ([action isEqualToString:@"git_push"] || [action isEqualToString:@"git_fetch"]) {
        NSString *cmd = [action isEqualToString:@"git_push"] ? @"push" : @"fetch";
        int r = system([[NSString stringWithFormat:@"git -C \"%@\" %@", node.path, cmd] UTF8String]);
        return @{ @"ok": @(r == 0) };
    }
    if ([action isEqualToString:@"fetch_merge"]) {
        NSString *remote = payload[@"remote"] ?: @"origin";
        NSString *branch = payload[@"branch"] ?: @"main";
        NSString *cmd = [NSString stringWithFormat:@"git -C \"%@\" fetch %@ && git -C \"%@\" merge %@/%@", node.path, remote, node.path, remote, branch];
        int r = system([cmd UTF8String]);
        return @{ @"ok": @(r == 0) };
    }
    if ([action isEqualToString:@"push_merge"]) {
        NSString *remote = payload[@"remote"] ?: @"origin";
        NSString *branch = payload[@"branch"] ?: @"main";
        NSString *cmd = [NSString stringWithFormat:@"(cd \"%@\" && git merge %@ 2>/dev/null; git push %@)", node.path, branch, remote];
        int r = system([cmd UTF8String]);
        return @{ @"ok": @(r == 0) };
    }
    if ([action isEqualToString:@"pull_merge"]) {
        NSString *remote = payload[@"remote"] ?: @"origin";
        NSString *branch = payload[@"branch"] ?: @"main";
        NSString *cmd = [NSString stringWithFormat:@"git -C \"%@\" pull %@ %@", node.path, remote, branch];
        int r = system([cmd UTF8String]);
        return @{ @"ok": @(r == 0) };
    }
    if ([action isEqualToString:@"job_submit"]) {
        NSString *script = payload[@"script"] ?: @"job.sh";
        NSString *cmd = [NSString stringWithFormat:@"cd \"%@\" && sbatch \"%@\"", node.path, script];
        int r = system([cmd UTF8String]);
        return @{ @"ok": @(r == 0) };
    }
    if ([action isEqualToString:@"notify"]) {
        return @{ @"ok": @(YES), @"message": @"notify not implemented" };
    }
    return @{ @"ok": @(NO), @"error": [NSString stringWithFormat:@"unknown action: %@", action] };
}
@end

File: blooming_directed_graph_agents/trigger_process_agent/swift/TriggerProcessAgent.swift
// blooming_directed_graph_trigger_process_agent – Swift.
// Namespace: LangFlow (flow/trigger).
// IEEE 752 64-bit geometrica Z field (zed): see docs/IEEE_752_64BIT_GEOMETRICA_Z_FIELD_ZED_SPECIFICATIONS.md

import Foundation

/// LangFlow: trigger node for process agent.
public struct TriggerNode {
    public var name: String
    public var path: String
    public var owner: String?
    public var repo: String?
    public var slug: String
}

public struct TriggerResult {
    public var ok: Bool
    public var error: String?
}

/// LangFlow namespace: trigger process agent.
public enum BloomingDirectedGraphTriggerProcessAgent {
    public static func trigger(node: TriggerNode, action: String, payload: [String: Any] = [:]) -> TriggerResult {
        switch action {
        case "workflow_dispatch":
            return triggerWorkflowDispatch(node: node, payload: payload)
        case "git_push":
            return triggerGit(node: node, cmd: "push")
        case "git_fetch":
            return triggerGit(node: node, cmd: "fetch")
        case "fetch_merge":
            return triggerFetchMerge(node: node, payload: payload)
        case "push_merge":
            return triggerPushMerge(node: node, payload: payload)
        case "pull_merge":
            return triggerGit(node: node, cmd: "pull")
        case "job_submit":
            return triggerHpccJob(node: node, payload: payload)
        case "run_script", "notify":
            return TriggerResult(ok: true, error: nil)
        default:
            return TriggerResult(ok: false, error: "unknown action: \(action)")
        }
    }

    static func triggerWorkflowDispatch(node: TriggerNode, payload: [String: Any]) -> TriggerResult {
        guard let _ = ProcessInfo.processInfo.environment["GITHUB_TOKEN"] else {
            return TriggerResult(ok: false, error: "GITHUB_TOKEN not set")
        }
        // Full implementation would use URLSession to POST to GitHub API
        return TriggerResult(ok: true, error: nil)
    }

    static func triggerGit(node: TriggerNode, cmd: String) -> TriggerResult {
        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/usr/bin/git")
        process.arguments = ["-C", node.path, cmd]
        do {
            try process.run()
            process.waitUntilExit()
            return TriggerResult(ok: process.terminationStatus == 0, error: nil)
        } catch {
            return TriggerResult(ok: false, error: error.localizedDescription)
        }
    }

    static func triggerFetchMerge(node: TriggerNode, payload: [String: Any]) -> TriggerResult {
        let remote = payload["remote"] as? String ?? "origin"
        let branch = payload["branch"] as? String ?? "main"
        let process1 = Process()
        process1.executableURL = URL(fileURLWithPath: "/usr/bin/git")
        process1.arguments = ["-C", node.path, "fetch", remote]
        do {
            try process1.run()
            process1.waitUntilExit()
            if process1.terminationStatus != 0 { return TriggerResult(ok: false, error: nil) }
        } catch { return TriggerResult(ok: false, error: error.localizedDescription) }
        let process2 = Process()
        process2.executableURL = URL(fileURLWithPath: "/usr/bin/git")
        process2.arguments = ["-C", node.path, "merge", "\(remote)/\(branch)"]
        do {
            try process2.run()
            process2.waitUntilExit()
            return TriggerResult(ok: process2.terminationStatus == 0, error: nil)
        } catch {
            return TriggerResult(ok: false, error: error.localizedDescription)
        }
    }

    static func triggerPushMerge(node: TriggerNode, payload: [String: Any]) -> TriggerResult {
        let remote = payload["remote"] as? String ?? "origin"
        let branch = payload["branch"] as? String ?? "main"
        let process1 = Process()
        process1.executableURL = URL(fileURLWithPath: "/usr/bin/git")
        process1.arguments = ["-C", node.path, "merge", branch]
        try? process1.run()
        process1.waitUntilExit()
        let process2 = Process()
        process2.executableURL = URL(fileURLWithPath: "/usr/bin/git")
        process2.arguments = ["-C", node.path, "push", remote]
        do {
            try process2.run()
            process2.waitUntilExit()
            return TriggerResult(ok: process2.terminationStatus == 0, error: nil)
        } catch {
            return TriggerResult(ok: false, error: error.localizedDescription)
        }
    }

    static func triggerHpccJob(node: TriggerNode, payload: [String: Any]) -> TriggerResult {
        let script = payload["script"] as? String ?? "job.sh"
        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/usr/bin/sbatch")
        process.arguments = [URL(fileURLWithPath: node.path).appendingPathComponent(script).path]
        process.currentDirectoryURL = URL(fileURLWithPath: node.path)
        do {
            try process.run()
            process.waitUntilExit()
            return TriggerResult(ok: process.terminationStatus == 0, error: nil)
        } catch {
            return TriggerResult(ok: false, error: error.localizedDescription)
        }
    }
}
